runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -457,30 +477,30 @@ public void insert(@ConflictParameter K key, V value) {
      * This method has a memory/CPU advantage over the map iterators as no deep copy
      * is actually performed.
      *
-     * @param p java predicate (function to evaluate)
+     * @param valuePredicate java predicate (function to evaluate)
      * @return a view of the values contained in this map meeting the predicate condition.
      */
     @Accessor
-    public List<V> scanAndFilter(Predicate<? super V> p) {
-        return mainMap.values().parallelStream()
-                                    .filter(p)
-                                    .collect(Collectors.toCollection(ArrayList::new));
+    public List<V> scanAndFilter(Predicate<? super V> valuePredicate) {
+        return mainMap.entryStream()
+                .map(Entry::getValue).filter(valuePredicate)
+                .collect(Collectors.toCollection(ArrayList::new));
     }
 
     /** {@inheritDoc} */
     @Override
     @Accessor
-    public Collection<Map.Entry<K, V>> scanAndFilterByEntry(Predicate<? super Map.Entry<K, V>>
-                                                                    entryPredicate) {
-        return mainMap.entrySet().parallelStream()
-                                    .filter(entryPredicate)
-                                    .collect(Collectors.toCollection(ArrayList::new));
+    public Collection<Map.Entry<K, V>> scanAndFilterByEntry(
+            Predicate<? super Map.Entry<K, V>> entryPredicate) {
+        return mainMap.entryStream()
+                .filter(entryPredicate)
+                .collect(Collectors.toCollection(ArrayList::new));
```

Comment:
`entryStream` can return a parallel stream and it will execute on the common pool, which is a forkjoin pool. This can cause two issues: 
1. If the threads block on io (via rocksdb), i think the forkjoin pool will spawn new threads to maintain "parallelism" 
2. If the stream is too large and since the forkjoin starts with `Runtime.getRuntime().availableProcessors() - 1 ` this might cause CPU starvation issues

I think better thread management is required here, or at least have this documented.

Reply:
The disk-backed CorfuTable always has to be single threaded. The actual behavior is defined in the mapImp. 

For memory (parallel):
```
    public Stream<Entry<K, V>> entryStream() {
        return mapImpl.entrySet().stream().parallel();
    }
```

For disk (single threaded):
```
    public Stream<Entry<K, V>> entryStream() {
        final RocksIterator rocksIterator = rocksDb.newIterator();
        rocksIterator.seekToFirst();
        return Streams.stream(new RocksDbIterator(rocksIterator));
    }
```

RocksDB itself does not allow multi-threaded access either.